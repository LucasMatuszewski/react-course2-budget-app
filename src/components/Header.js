import React from 'react';
import { BrowserRouter, Route, Switch, Link, NavLink } from 'react-router-dom';

const Header = () => (
    <div>
        <h1>Expenses App</h1>
        {/*
        NavLink is for navigation menu ( adds classes/style to active links). Normal link is <Link>
        For main page "/" use exact like in <Route /> - other way it will match all sites
        You could use ActiveClassName or activeStyle to style inline (syntax for CSS is camelCase like for HTML tags in JSX)
        We can use also other HTML props like title="" or id="", className=""
        */}
        <NavLink to="/" exact activeClassName="is-active" activeStyle={{ fontWeight: 'bold', color: 'red' }} title="test">Dashboard</NavLink> | <NavLink to="/create" activeClassName="is-active" activeStyle={{ fontWeight: 'bold', color: 'red' }} title="test">Add Expense</NavLink>
        | <NavLink to="/help" activeClassName="is-active" activeStyle={{ fontWeight: 'bold', color: 'red' }} title="test">Help</NavLink><br /> {/* like in XHTML/XML we have to end tags with "/" */}
        <hr />
    </div>
);

export default Header;