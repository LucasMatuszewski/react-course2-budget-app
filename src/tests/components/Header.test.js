/****************************
 * REACT-TEST-RENDER version:
 ****************************/

/* 
// we have to virtually render this component, to figure out what is rendered.
// But we won't do it in a browser. We have to install React Library:
// > yarn add react-test-renderer -D

import React from 'react';
import ReactShallowRenderer from 'react-test-renderer/shallow';
import Header from '../../components/Header';

test('should render Header correctly', () => {
    const renderer = new ReactShallowRenderer();
    renderer.render(<Header />);
    console.log(renderer.getRenderOutput()); // we can console.log to JEST console.
    // But we will not compare it manually ;)
    // We use snapshots to automatically track changes to data over time.
    // We create a snapshot of header in current point of time and we get notify when it change.
    // Jest method: .toMatchSnapshot(optionalString)
    // https://facebook.github.io/jest/docs/en/expect.html#tomatchsnapshotpropertymatchers-snapshotname
    // https://facebook.github.io/jest/docs/en/snapshot-testing

    expect(renderer.getRenderOutput()).toMatchSnapshot();
    // First time we run this test it will pass, because there is no existing snapshot.
    // It take a snapshot of rendered UI and track changes compering to this snapshot.
    // snapshots are saved in tests/components/__snapshots__ - we can read / review it.

    // With shallow rendering we don't see <Link> anchor tags in test.js.snap files
});
*/

/*****************
 * ENZYME Version:
 * much easier & user friendly API
 *********************************/

import React from 'react';
import { shallow } from 'enzyme';
import Header from '../../components/Header';

test('should render Header correctly', () => {
    const wrapper = shallow(<Header />);
    // Now we can use Enzyme methods on wrapper (the syntax is very similar to jQuery)
    // e.g. .find('selector') to return array with founded elements
    // selector could be CSS Selector (Element / Class / ID) or Component, variable, Object prop
    expect(wrapper.find('h1').length).toBe(1); // FOR SEO
    // .length = number of elements in array/object = if there is only one <h1> found = 1
    expect(wrapper.find('h1').text()).toBe('Expenses App'); // to check a text of h1
    // http://airbnb.io/enzyme/docs/api/ShallowWrapper/text.html

    expect(wrapper).toMatchSnapshot();
    // By default snapshot from Enzyme is full of Enzyme properties, we don't need.
    // We have to use enzyme-to-json library tu make snapshots with Enzyme
});