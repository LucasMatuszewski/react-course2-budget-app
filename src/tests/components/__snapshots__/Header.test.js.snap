// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Header correctly 1`] = `
<div>
  <h1>
    Expenses App
  </h1>
  <NavLink
    activeClassName="is-active"
    activeStyle={
      Object {
        "color": "red",
        "fontWeight": "bold",
      }
    }
    aria-current="page"
    exact={true}
    title="test"
    to="/"
  >
    Dashboard
  </NavLink>
   | 
  <NavLink
    activeClassName="is-active"
    activeStyle={
      Object {
        "color": "red",
        "fontWeight": "bold",
      }
    }
    aria-current="page"
    title="test"
    to="/create"
  >
    Add Expense
  </NavLink>
  | 
  <NavLink
    activeClassName="is-active"
    activeStyle={
      Object {
        "color": "red",
        "fontWeight": "bold",
      }
    }
    aria-current="page"
    title="test"
    to="/help"
  >
    Help
  </NavLink>
  <br />
   
  <hr />
</div>
`;
